File: python/src/starbridge/protocol.py
from __future__ import annotations

from dataclasses import dataclass
import struct

from .constants import FrameKind, MAX_PAYLOAD, PROTOCOL_VERSION
from .errors import ProtocolError


SOF = b"\xA5\x5A"
_HEADER = struct.Struct("<BBHBH")
_CRC = struct.Struct("<H")
MIN_FRAME_SIZE = len(SOF) + _HEADER.size + _CRC.size


def crc16_ccitt(data: bytes, initial: int = 0xFFFF) -> int:
    crc = initial
    for byte in data:
        crc ^= byte << 8
        for _ in range(8):
            crc = ((crc << 1) ^ 0x1021) & 0xFFFF if crc & 0x8000 else (crc << 1) & 0xFFFF
    return crc


@dataclass(frozen=True, slots=True)
class Frame:
    kind: FrameKind
    sequence: int
    command: int
    payload: bytes = b""
    version: int = PROTOCOL_VERSION

    def encode(self) -> bytes:
        if not 0 <= self.sequence <= 0xFFFF:
            raise ValueError("sequence must fit in uint16")
        if not 0 <= self.command <= 0xFF:
            raise ValueError("command must fit in uint8")
        if len(self.payload) > MAX_PAYLOAD:
            raise ValueError(f"payload exceeds {MAX_PAYLOAD} bytes")
        body = _HEADER.pack(
            self.version,
            int(self.kind),
            self.sequence,
            self.command,
            len(self.payload),
        ) + self.payload
        return SOF + body + _CRC.pack(crc16_ccitt(body))


class StreamDecoder:
    """Incremental decoder that resynchronizes after noise or corrupt frames."""

    def __init__(self) -> None:
        self._buffer = bytearray()
        self.bad_crc_count = 0
        self.bad_length_count = 0

    def reset(self) -> None:
        self._buffer.clear()

    def feed(self, data: bytes) -> list[Frame]:
        self._buffer.extend(data)
        frames: list[Frame] = []
        while True:
            start = self._buffer.find(SOF)
            if start < 0:
                if self._buffer[-1:] == SOF[:1]:
                    del self._buffer[:-1]
                else:
                    self._buffer.clear()
                break
            if start:
                del self._buffer[:start]
            if len(self._buffer) < MIN_FRAME_SIZE:
                break
            version, raw_kind, sequence, command, length = _HEADER.unpack_from(self._buffer, len(SOF))
            if length > MAX_PAYLOAD:
                self.bad_length_count += 1
                del self._buffer[0]
                continue
            frame_size = MIN_FRAME_SIZE + length
            if len(self._buffer) < frame_size:
                break
            body_end = len(SOF) + _HEADER.size + length
            body = bytes(self._buffer[len(SOF):body_end])
            expected_crc = _CRC.unpack_from(self._buffer, body_end)[0]
            if crc16_ccitt(body) != expected_crc:
                self.bad_crc_count += 1
                del self._buffer[0]
                continue
            try:
                kind = FrameKind(raw_kind)
            except ValueError:
                del self._buffer[:frame_size]
                continue
            frames.append(Frame(kind, sequence, command, body[_HEADER.size:], version))
            del self._buffer[:frame_size]
        return frames


def decode_frame(data: bytes) -> Frame:
    decoder = StreamDecoder()
    frames = decoder.feed(data)
    if len(frames) != 1 or decoder._buffer:
        raise ProtocolError("expected exactly one complete frame")
    return frames[0]

File: firmware/starcore-v2/src/protocol/Protocol.cpp
#include "Protocol.h"

#include <cstring>

namespace protocol {

uint16_t crc16Ccitt(const uint8_t* data, size_t length, uint16_t initial) {
  uint16_t crc = initial;
  for (size_t index = 0; index < length; ++index) {
    crc ^= static_cast<uint16_t>(data[index]) << 8;
    for (uint8_t bit = 0; bit < 8; ++bit) {
      crc = (crc & 0x8000) ? static_cast<uint16_t>((crc << 1) ^ 0x1021)
                           : static_cast<uint16_t>(crc << 1);
    }
  }
  return crc;
}

void Decoder::reset() {
  state_ = State::SeekFirst;
  received_ = 0;
  expected_ = 0;
}

void Decoder::tick(uint32_t nowMs) {
  if (state_ != State::SeekFirst && static_cast<uint32_t>(nowMs - lastByteMs_) > FRAME_TIMEOUT_MS) {
    reset();
  }
}

bool Decoder::feed(uint8_t byte, uint32_t nowMs, Frame& output) {
  lastByteMs_ = nowMs;
  if (state_ == State::SeekFirst) {
    if (byte == SOF_0) state_ = State::SeekSecond;
    return false;
  }
  if (state_ == State::SeekSecond) {
    if (byte == SOF_1) {
      state_ = State::Body;
      received_ = 0;
      expected_ = 0;
    } else {
      state_ = byte == SOF_0 ? State::SeekSecond : State::SeekFirst;
    }
    return false;
  }

  body_[received_++] = byte;
  if (received_ == HEADER_SIZE) {
    const uint16_t payloadLength = static_cast<uint16_t>(body_[5]) |
                                   (static_cast<uint16_t>(body_[6]) << 8);
    if (payloadLength > MAX_PAYLOAD) {
      ++lengthErrors_;
      reset();
      return false;
    }
    expected_ = HEADER_SIZE + payloadLength + 2;
  }
  if (expected_ == 0 || received_ < expected_) return false;

  const uint16_t receivedCrc = static_cast<uint16_t>(body_[expected_ - 2]) |
                               (static_cast<uint16_t>(body_[expected_ - 1]) << 8);
  const uint16_t calculatedCrc = crc16Ccitt(body_, expected_ - 2);
  if (receivedCrc != calculatedCrc) {
    ++crcErrors_;
    reset();
    return false;
  }

  output.version = body_[0];
  output.kind = static_cast<Kind>(body_[1]);
  output.sequence = static_cast<uint16_t>(body_[2]) |
                    (static_cast<uint16_t>(body_[3]) << 8);
  output.command = body_[4];
  output.length = static_cast<uint16_t>(body_[5]) |
                  (static_cast<uint16_t>(body_[6]) << 8);
  if (output.length) std::memcpy(output.payload, body_ + HEADER_SIZE, output.length);
  reset();
  return true;
}

size_t encode(const Frame& frame, uint8_t* output, size_t capacity) {
  const size_t required = 2 + HEADER_SIZE + frame.length + 2;
  if (frame.length > MAX_PAYLOAD || capacity < required) return 0;
  output[0] = SOF_0;
  output[1] = SOF_1;
  output[2] = frame.version;
  output[3] = static_cast<uint8_t>(frame.kind);
  output[4] = static_cast<uint8_t>(frame.sequence & 0xFF);
  output[5] = static_cast<uint8_t>(frame.sequence >> 8);
  output[6] = frame.command;
  output[7] = static_cast<uint8_t>(frame.length & 0xFF);
  output[8] = static_cast<uint8_t>(frame.length >> 8);
  if (frame.length) std::memcpy(output + 9, frame.payload, frame.length);
  const uint16_t crc = crc16Ccitt(output + 2, HEADER_SIZE + frame.length);
  output[9 + frame.length] = static_cast<uint8_t>(crc & 0xFF);
  output[10 + frame.length] = static_cast<uint8_t>(crc >> 8);
  return required;
}

}  // namespace protocol

